Add a test to ensure DDP join works with RPC

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44439 Add a test to ensure DDP join works with RPC**

Adds a test to ddp_under_dist_autograd_test to enusre that that uneven
inputs join() API works properly when DDP + RPC is combined. We test that when
running in outside DDP mode (DDP applied to whole hybrid module) we can
correctly process uneven inputs across different trainers.

Differential Revision: [D23612409](https://our.internmc.facebook.com/intern/diff/D23612409/)

rohan-varma committed:
Add a test to ensure DDP join works with RPC

Adds a test to ddp_under_dist_autograd_test to enusre that that uneven
inputs join() API works properly when DDP + RPC is combined. We test that when
running in outside DDP mode (DDP applied to whole hybrid module) we can
correctly process uneven inputs across different trainers.

Differential Revision: [D23612409](https://our.internmc.facebook.com/intern/diff/D23612409/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add a test to ensure DDP join works with RPC"


Adds a test to ddp_under_dist_autograd_test to enusre that that uneven
inputs join() API works properly when DDP + RPC is combined. We test that when
running in outside DDP mode (DDP applied to whole hybrid module) we can
correctly process uneven inputs across different trainers.

Differential Revision: [D23612409](https://our.internmc.facebook.com/intern/diff/D23612409/)

[ghstack-poisoned]

pritamdamania87 commented on torch/testing/_internal/distributed/ddp_under_dist_autograd_test.py:
```diff
@@ -218,17 +219,50 @@ def destroy_pg(self):
+            else:
+                input_batches = batches
+
+        with self.hybrid_module.join() if simulate_uneven_inputs else contextlib.suppress():
```
nit: you can just pass `simulate_uneven_inputs` to the `enable` param of `join()`?

rohan-varma replied:
Actually, we probably have to stick with `suppress()` since `self.hybrid_module` is not always an instance of DDP in some of the test cases.

pritamdamania87 commented on torch/testing/_internal/distributed/ddp_under_dist_autograd_test.py:
```diff
@@ -366,24 +406,35 @@ def do_test_on_master(
-                    )
+                # When there are uneven inputs, it is not necessary that grads
+                # cancel each other out, since some trainers contribute 0 grad.
+                if not simulate_uneven_inputs:
+                    for grad in ddp_grads:
+                        self.assertEqual(
+                            grad,
+                            torch.zeros_like(grad),
+                            msg=f"The grad for any ddp parameter should be zeros, because "
+                            "the training examples' grads cancel each other. Received "
+                            f"gradient {grad}",
+                        )
```
We should probably still add a test that at least validates the grads. Maybe add a test under DdpComparisonTest?

rohan-varma replied:
Added a test for this by varying `test_ddp_comparison` to allow uneven inputs.

rohan-varma committed:
Update on "Add a test to ensure DDP join works with RPC"


Adds a test to ddp_under_dist_autograd_test to enusre that that uneven
inputs join() API works properly when DDP + RPC is combined. We test that when
running in outside DDP mode (DDP applied to whole hybrid module) we can
correctly process uneven inputs across different trainers.

Differential Revision: [D23612409](https://our.internmc.facebook.com/intern/diff/D23612409/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add a test to ensure DDP join works with RPC"


Adds a test to ddp_under_dist_autograd_test to enusre that that uneven
inputs join() API works properly when DDP + RPC is combined. We test that when
running in outside DDP mode (DDP applied to whole hybrid module) we can
correctly process uneven inputs across different trainers.

Differential Revision: [D23612409](https://our.internmc.facebook.com/intern/diff/D23612409/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add a test to ensure DDP join works with RPC"


Adds a test to ddp_under_dist_autograd_test to enusre that that uneven
inputs join() API works properly when DDP + RPC is combined. We test that when
running in outside DDP mode (DDP applied to whole hybrid module) we can
correctly process uneven inputs across different trainers.

Differential Revision: [D23612409](https://our.internmc.facebook.com/intern/diff/D23612409/)

[ghstack-poisoned]
